source/extensions/filters/network/dubbo_proxy/router/route_matcher.h

```diff
@@ -0,0 +1,162 @@
+  bool headersMatch(const Http::HeaderMap& headers) const;
+
+private:
+  class WeightedClusterEntry : public RouteEntry, public Route {
+  public:
+    using WeightedCluster = envoy::api::v2::route::WeightedCluster_ClusterWeight;
+    WeightedClusterEntry(const RouteEntryImplBase& parent, const WeightedCluster& cluster);
+
+    uint64_t clusterWeight() const { return cluster_weight_; }
+
+    // Router::RouteEntry
+    const std::string& clusterName() const override { return cluster_name_; }
+    const Envoy::Router::MetadataMatchCriteria* metadataMatchCriteria() const override {
+      return metadata_match_criteria_ ? metadata_match_criteria_.get()
+                                      : parent_.metadataMatchCriteria();
+    }
+
+    // Router::Route
+    const RouteEntry* routeEntry() const override { return this; }
+
+  private:
+    const RouteEntryImplBase& parent_;
+    const std::string cluster_name_;
+    const uint64_t cluster_weight_;
+    Envoy::Router::MetadataMatchCriteriaConstPtr metadata_match_criteria_;
+  };
+
+  typedef std::shared_ptr<WeightedClusterEntry> WeightedClusterEntrySharedPtr;
+
+  uint64_t total_cluster_weight_;
+  const std::string cluster_name_;
+  std::vector<Http::HeaderUtility::HeaderData> config_headers_;
+  std::vector<WeightedClusterEntrySharedPtr> weighted_clusters_;
+
+  // TODO(leilei.gll) Implement it.
+  Envoy::Router::MetadataMatchCriteriaConstPtr metadata_match_criteria_;
+};
+
+typedef std::shared_ptr<const RouteEntryImplBase> RouteEntryImplBaseConstSharedPtr;
+
+class ParameterRouteEntryImpl : public RouteEntryImplBase {
+public:
+  ParameterRouteEntryImpl(
+      const envoy::config::filter::network::dubbo_proxy::v2alpha1::Route& route);
+  ~ParameterRouteEntryImpl() override;
+
+  struct ParameterData {
+    using ParameterMatchSpecifier =
+        envoy::config::filter::network::dubbo_proxy::v2alpha1::MethodMatch_ParameterMatchSpecifier;
+    ParameterData(uint32_t index, const ParameterMatchSpecifier& config);
+
+    Http::HeaderUtility::HeaderMatchType match_type_;
+    std::string value_;
+    envoy::type::Int64Range range_;
+    uint32_t index_;
+  };
+
+  // RoutEntryImplBase
+  RouteConstSharedPtr matches(const MessageMetadata& metadata,
+                              uint64_t random_value) const override;
+
+private:
+  bool matchParameter(const std::string& request_data, const ParameterData& config_data) const;
+
+  std::vector<ParameterData> parameter_data_list_;
+};
+
+class MethodRouteEntryImpl : public RouteEntryImplBase {
+public:
+  MethodRouteEntryImpl(const envoy::config::filter::network::dubbo_proxy::v2alpha1::Route& route);
+  ~MethodRouteEntryImpl() override;
+
+  // RoutEntryImplBase
+  RouteConstSharedPtr matches(const MessageMetadata& metadata,
+                              uint64_t random_value) const override;
+
+private:
+  const Matchers::StringMatcher method_name_;
+  absl::optional<std::shared_ptr<ParameterRouteEntryImpl>> parameter_route_;
```

Comment:
do you need to differentiate nullopt vs nullptr? If no simple shared_ptr should be enough.

Reply:
done